### ISSUE TITLE
Qt6, PyQt6: v0.4.17rc5 viewer closing on object call in a Jupyter kernel

### ISSUE THREAD

**Issue by Tomaubier** [napari/napari#5273]
## 🐛 Bug

Napari viewer v0.4.17rc5 crashes when calling the `viewer` object in a Jupyter notebook (tested in VS Code) or Jupyter console but works fine in `IPython` as found by @psobolewskiPhD in #3880.

## To Reproduce

Steps to reproduce in a terminal:
1. Launch a Jupiter kernel (I used `jupyter console`)
2. `In [1]: import napari`
3. `In [2]: viewer = napari.Viewer()` -> the viewer opens and behaves as expected (other than the graphical bug mentioned by @tlambert03 in  #3880).
4. `In [3]: <run any python command>` (`print('foo')` for example) -> the viewer window closes but the viewer object remains accessible (calling `viewer.dims works, however adding a layer (eg `viewer.add_shapes()`) throws the following error).

```
In [5]: viewer.add_shapes()
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In [5], line 1
----> 1 viewer.add_shapes()

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/components/viewer_model.py:5, in add_shapes(self, data, ndim, features, properties, property_choices, text, shape_type, edge_width, edge_color, edge_color_cycle, edge_colormap, edge_contrast_limits, face_color, face_color_cycle, face_colormap, face_contrast_limits, z_index, name, metadata, scale, translate, rotate, shear, affine, opacity, blending, visible, cache, experimental_clipping_planes)
      3 import inspect
      4 import itertools
----> 5 import os
      6 import warnings
      7 from functools import lru_cache

File ~/miniconda3/envs/neustim310/lib/python3.10/_collections_abc.py:1123, in MutableSequence.append(self, value)
   1121 def append(self, value):
   1122     'S.append(value) -- append value to the end of the sequence'
-> 1123     self.insert(len(self), value)

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/components/layerlist.py:166, in LayerList.insert(self, index, value)
    164 self._clean_cache()
    165 new_layer.events.set_data.connect(self._clean_cache)
--> 166 super().insert(index, new_layer)

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/utils/events/containers/_selectable_list.py:69, in SelectableEventedList.insert(self, index, value)
     66 super().insert(index, value)
     67 if self._activate_on_insert:
     68     # Make layer selected and unselect all others
---> 69     self.selection.active = value

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/utils/events/containers/_selection.py:104, in Selection.active(self, value)
    102 self.clear() if value is None else self.select_only(value)
    103 self._current = value
--> 104 self.events.active(value=value)

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/utils/events/event.py:757, in EventEmitter.__call__(self, *args, **kwargs)
    754     self._block_counter.update([cb])
    755     continue
--> 757 self._invoke_callback(cb, event if pass_event else None)
    758 if event.blocked:
    759     break

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/utils/events/event.py:794, in EventEmitter._invoke_callback(self, cb, event)
    792     self.disconnect(cb)
    793     return
--> 794 _handle_exception(
    795     self.ignore_callback_errors,
    796     self.print_callback_errors,
    797     self,
    798     cb_event=(cb, event),
    799 )

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/utils/events/event.py:781, in EventEmitter._invoke_callback(self, cb, event)
    779 try:
    780     if event is not None:
--> 781         cb(event)
    782     else:
    783         cb()

File ~/miniconda3/envs/neustim310/lib/python3.10/site-packages/napari/_qt/layer_controls/qt_layer_controls_container.py:116, in QtLayerControlsContainer._display(self, event)
    114     self.setCurrentWidget(self.empty_widget)
    115 else:
--> 116     controls = self.widgets[layer]
    117     self.setCurrentWidget(controls)

KeyError: <Shapes layer 'Shapes' at 0x2887a76d0>
```

## Expected behavior

The viewer window should remain opened and active when its object is called.

## Environment

```
napari: 0.4.17rc5
Platform: macOS-13.0-arm64-arm-64bit <- Also reproducible with macOS-12.5.1-arm64-arm-64bit
System: MacOS 13.0
Python: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ]
Qt: 6.4.0
PyQt6: 
NumPy: 1.23.4
SciPy: 1.9.3
Dask: 2022.10.0
VisPy: 0.11.0
magicgui: 0.6.0rc1
superqt: 0.3.8
in-n-out: 0.1.5
app-model: 0.1.0
npe2: 0.6.1

OpenGL:
- GL version: 2.1 Metal - 83
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1800x1169, scale 2.0
- screen 2: resolution 1920x1080, scale 1.0
```

## Additional context

The issue can't be reproduced using the `IPython`console:
```
% ipython
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import napari
In [2]: viewer = napari.Viewer()
In [3]: viewer.add_shapes()
Out[3]: <Shapes layer 'Shapes' at 0x1067eb910>
```
-> the viewer stays open and usable

**Comment by andy-sweet** [comment]
Thanks for testing the RC and reporting the bug @Tomaubier!

Qt6 support is still experimental for v0.4.17, so this probably won't be fixed with the final release for v0.4.17. But given its severeness, we should try to prioritize this to get into the next release.

**Comment by Tomaubier** [comment]
Thanks for the reply!
To my knowledge and from my experience Qt6 is the only version of Qt running natively on Apple silicon. Fixing this quite severe bug significantly reducing the usability of Napari on this platform seems like a good idea!
Especially since, from what I’ve tested using IPython, most of Napari’s features seem stable under Qt6.

**Comment by Czaki** [comment]
@Tomaubier I understand, but this requires someone to spend time on it and we do not want to delay release of napari. 

If you have time to try debug this and prepare a simple bugfix it is possible to add this to 0.4.17 or maybe 0.4.18. 

In the opposite situation you need to wait until someone will have a free time for bugfixing napari.

**Comment by psobolewskiPhD** [comment]
> Thanks for the reply!
> To my knowledge and from my experience Qt6 is the only version of Qt running natively on Apple silicon. Fixing this quite severe bug significantly reducing the usability of Napari on this platform seems like a good idea!
> Especially since, from what I’ve tested using IPython, most of Napari’s features seem stable under Qt6.

@Tomaubier if you want to run arm64 native on M1 you can use pyqt5 from conda-forge.
Make sure your conda env is arm64, I use miniforge3:
https://github.com/conda-forge/miniforge
I also use mamba.
Then you can create an env with pyqt5 from conda-forge and `pip install napari` plus whatever
or you can install napari from conda-forge directly and be done.

Edit: but yes, on pypi it's just pyqt5 and pyside6 that are native arm64 wheels.
Edit2: vispy just had a PR merged to fix pyside6, so once that's released I wonder if that's a way to work around this problem. I will test with `main` vispy.
Edit: alas, pyside6 doesn't work due to something else, in app-model. I will poke around some more and make a new issue.

**Comment by psobolewskiPhD** [comment]
So with pyside6 6.3.2 and rc6 I get the same behavior as the OP, so it's not a PyQt issue, but Qt6 in general.
Note that after the viewer disappears, I tried `viewer.show()` and got:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
File ~/Dev/miniforge3/envs/nap-test/lib/python3.10/site-packages/napari/_qt/qt_main_window.py:1088, in Window.show(self, block)
   1087 try:
-> 1088     self._qt_window.show(block=block)
   1089 except (AttributeError, RuntimeError):

File ~/Dev/miniforge3/envs/nap-test/lib/python3.10/site-packages/napari/_qt/qt_main_window.py:327, in _QtMainWindow.show(self, block)
    326 def show(self, block=False):
--> 327     super().show()
    328     self._qt_viewer.setFocus()

RuntimeError: Internal C++ object (_QtMainWindow) already deleted.

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
Cell In [6], line 1
----> 1 viewer.show()

File ~/Dev/miniforge3/envs/nap-test/lib/python3.10/site-packages/napari/viewer.py:139, in Viewer.show(self, block)
    137 def show(self, *, block=False):
    138     """Resize, show, and raise the viewer window."""
--> 139     self.window.show(block=block)

File ~/Dev/miniforge3/envs/nap-test/lib/python3.10/site-packages/napari/_qt/qt_main_window.py:1090, in Window.show(self, block)
   1088     self._qt_window.show(block=block)
   1089 except (AttributeError, RuntimeError):
-> 1090     raise RuntimeError(
   1091         trans._(
   1092             "This viewer has already been closed and deleted. Please create a new one.",
   1093             deferred=True,
   1094         )
   1095     )
   1097 if settings.application.first_time:
   1098     settings.application.first_time = False

RuntimeError: This viewer has already been closed and deleted. Please create a new one.
```

**Comment by psobolewskiPhD** [comment]
This will be fixed with ipykernel 6.20 release:
https://github.com/ipython/ipykernel/pull/1054
was yanked for some regressions, but you can test with 
pip install git+https://github.com/ipython/ipykernel.git

**Comment by psobolewskiPhD** [comment]
With the latest ipykernel release, this has been fixed.
Fresh 3.10 env
`pip install pyqt6`
(6.4.2)
`pip install napari`
(0.4.17)
This yields ipykernel 6.20.1 and in both jupyter qtconsole and VSCode jupyter notebook the issue is resolved.

Note: pyside6 at the moment has other issues: https://github.com/jupyter/qtconsole/issues/567 and https://github.com/napari/napari/issues/5479
